src/Widgets/calendar/GridDay.vala

```diff
@@ -84,53 +82,49 @@ public class DateTime.Widgets.GridDay : Gtk.EventBox {
-        foreach (var component in events) {
-            if (event_dots.size >= 3) {
-                return;
-            }
 
-            if (Util.calcomp_is_on_day (component, date)) {
-                unowned ICal.Component ical = component.get_icalcomponent ();
+    public void add_dots (E.Source source, ICal.Component ical) {
+        var event_uid = ical.get_uid ();
+        if (event_dots.contains (event_uid)) {
+            return;
+        }
 
-                var event_uid = ical.get_uid ();
-                if (!event_dots.has_key (event_uid)) {
-                    var event_dot = new Gtk.Image ();
-                    event_dot.gicon = new ThemedIcon ("pager-checked-symbolic");
-                    event_dot.pixel_size = 6;
+        event_dots.add (event_uid);
+        if (event_dots.size > 3) {
+            return;
+        }
 
-                    unowned Gtk.StyleContext style_context = event_dot.get_style_context ();
-                    style_context.add_class (Granite.STYLE_CLASS_ACCENT);
-                    style_context.add_provider (provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
+        var event_dot = new Gtk.Image ();
+        event_dot.gicon = new ThemedIcon ("pager-checked-symbolic");
+        event_dot.pixel_size = 6;
 
-                    var source_calendar = (E.SourceCalendar?) source.get_extension (E.SOURCE_EXTENSION_CALENDAR);
-                    Util.set_event_calendar_color (source_calendar, event_dot);
+        unowned Gtk.StyleContext style_context = event_dot.get_style_context ();
+        style_context.add_class (Granite.STYLE_CLASS_ACCENT);
+        style_context.add_provider (provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
 
-                    event_dots[event_uid] = event_dot;
+        var source_calendar = (E.SourceCalendar?) source.get_extension (E.SOURCE_EXTENSION_CALENDAR);
+        Util.set_event_calendar_color (source_calendar, event_dot);
 
-                    event_grid.add (event_dot);
-                }
-            }
+        event_grid.add (event_dot);
+        event_dot.show ();
+    }
+
+    public void remove_dots (string event_uid) {
+        if (!event_dots.contains (event_uid)) {
+            return;
         }
 
-        event_grid.show_all ();
-    }
+        event_dots.remove (event_uid);
+        if (event_dots.size >= 3) {
+            return;
+        }
 
-    private void remove_event_dots (E.Source source, Gee.Collection<ECal.Component> events) {
-        foreach (var component in events) {
-            unowned ICal.Component ical = component.get_icalcomponent ();
-            var event_uid = ical.get_uid ();
-            var dot = event_dots[event_uid];
-            if (dot != null) {
-                dot.destroy ();
-                event_dots.remove (event_uid);
-            }
+        var dot = event_grid.get_children ();
+        if (dot.length () > 0) {
+            dot.nth_data (0).destroy ();
```

Comment:
This seems less efficient and less accurate than using the hashmap as we have it in master. There's no guarantee that the first dot is the correct one to remove, which means you could have dots with colors that don't match what's actually on that day

Reply:
Currently, all your points are depersonalized. Just point with a single class, why should they be personalized? If you try to add, say, 6 events, only 3 will be added. Now that you delete any of the added 3 events, what happens? correctly, the point will be deleted. As a result, you will get 2 points and 5 events.
The code that you reject at least checks whether it is time to delete a point, or only reduce the number of events for this point

Reply:
What's a better solution for this? Should I run an iterator, search the list and destroy the element that way?

Reply:
I outlined my opinion above. I hope someone from the developers will voice what logic they laid. I assume that I am wrong, because I do not use this functionality)

Reply:
@Dirli and @danrabbit. I think I understand what @danrabbit is saying. I still feel an arraylist is faster than the hashmap implementation. The problem is the dots are color coded for different types of events (I've never noticed this before). So, if we always remove the first dot, that might not be the right one to remove.

Reply:
@Dirli I'm not a developer so I'm struggling here. Is there a way to search the arraylist and remove an element from that list based on what event_uid is?

Reply:
Yes, the colors were a surprise to me (while fix issue, I did not come across them). My solution needs to be rewritten to work correctly. Leaving the discussion)